src/format/utils.rs

```diff
@@ -0,0 +1,101 @@
+//! Some useful types.
+
+use super::write::Write;
+use crate::Error;
+
+/// A `Cursor` contains a slice of a buffer.
+#[derive(Debug, Clone)]
+pub(crate) struct Cursor<'a> {
+    /// Slice representing the remaining data to be read.
+    remaining: &'a [u8],
+}
+
+impl<'a> Cursor<'a> {
+    /// Construct a new `Cursor` from remaining data.
+    pub(crate) fn new(remaining: &'a [u8]) -> Self {
+        Self { remaining }
+    }
+
+    /// Returns remaining data.
+    pub(crate) fn remaining(&self) -> &'a [u8] {
+        self.remaining
+    }
+
+    /// Returns the next byte.
+    pub(crate) fn next(&mut self) -> Option<u8> {
+        match self.remaining {
+            [first, tail @ ..] => {
+                self.remaining = tail;
+                Some(*first)
+            }
+            _ => None,
+        }
```

Comment:
optional style nit, we can eliminate some rightward drift by using `<[_]>::split_first`:

```suggestion
        let (&first, tail) = self.remaining.split_first()?;
        self.remaining = tail;
        Some(first)
```